tests/cartesian_power.rs

```diff
@@ -0,0 +1,486 @@
+                println!(" {act:?}");
+                // Possible failure report.
+                let fail = |e, a| {
+                    let f = |o| {
+                        if let Some(v) = o {
+                            format!("{v:?}")
+                        } else {
+                            "None".into()
+                        }
+                    };
+                    panic!("{} Expected {:?}, got {:?} instead.", ctx(), f(e), f(a));
+                };
+                // Comparison.
+                match (exp, act) {
+                    (Some(exp), Some(act)) => {
+                        if act != exp {
+                            fail(Some(exp), Some(act));
+                        }
+                        i += 1;
+                    }
+                    (None, Some(act)) => {
+                        fail(None, Some(act));
+                    }
+                    (Some(exp), None) => {
+                        fail(Some(exp), None);
+                    }
+                    (None, None) => break,
+                }
+                e_remaining -= 1;
+            }
+        }
+    }
+
+    // Empty underlying iterator.
+    check("", 0, &[""]); // 0^0 = 1.
+    check("", 1, &[]);
+    check("", 2, &[]);
+    check("", 3, &[]);
+
+    // Singleton underlying iterator.
+    check("a", 0, &[""]);
+    check("a", 1, &["a"]);
+    check("a", 2, &["aa"]);
+    check("a", 3, &["aaa"]);
+
+    // Underlying pair.
+    check("ab", 0, &[""]);
+    check("ab", 1, &["a", "b"]);
+    check("ab", 2, &["aa", "ab", "ba", "bb"]);
+    check(
+        "ab",
+        3,
+        &["aaa", "aab", "aba", "abb", "baa", "bab", "bba", "bbb"],
+    );
+
+    // Underlying triplet.
+    check("abc", 0, &[""]);
+    check("abc", 1, &["a", "b", "c"]);
+    check(
+        "abc",
+        2,
+        &["aa", "ab", "ac", "ba", "bb", "bc", "ca", "cb", "cc"],
+    );
+    check(
+        "abc",
+        3,
+        &[
+            "aaa", "aab", "aac", "aba", "abb", "abc", "aca", "acb", "acc", "baa", "bab", "bac",
+            "bba", "bbb", "bbc", "bca", "bcb", "bcc", "caa", "cab", "cac", "cba", "cbb", "cbc",
+            "cca", "ccb", "ccc",
+        ],
+    );
+}
+
+#[test]
+#[allow(clippy::too_many_lines)] // Numerous detailed tests, especially corner-cases.
+#[allow(clippy::similar_names)] // On-purpose `aab`, `aba`, `aac`, etc.
+#[allow(clippy::many_single_char_names)] // On-purpose short `o`, `e`, `a`: ease test readability.
+fn nth() {
+    fn check(origin: &str, pow: u32, expected: &[(usize, (Option<&str>, usize))]) {
+        println!("================== ({origin:?}^{pow})");
+        let mut it = origin
+            .chars()
+            .collect::<Vec<_>>()
+            .into_iter()
+            .cartesian_power(pow);
+        let mut total_n = Vec::new();
+        for &(n, (exp, e_hint)) in expected {
+            total_n.push(n);
+            let ctx = || {
```

Comment:
I don't think we usually use that detailed error reporting - it is not needed in most cases.

Reply:
Yup. I really was useful to me when debugging my implementation. But I can remove it now. Lemme clean that up.